flipperdevices/flipperzero-firmware
# Implement support for reading Opal card (Sydney, Australia)

# What's new

- Adds support for reading Opal card (Sydney, Australia): [format reference](https://github.com/metrodroid/metrodroid/wiki/Opal)
- Adds hooks for parsing MIFARE DESFire cards, based on MIFARE Classic and Ultralight readers. Unlike the others, it doesn't use `.verify` or `.read`, so we have a stub `stub_parser_verify_read` for that.
- Adds `mf_df_get_application` and `mf_df_get_file` helpers for parsing DESFire cards.
- Adds `furi_hal_rtc_get_days_per_month`, `furi_hal_rtc_get_days_per_year`, `furi_hal_rtc_is_leap_year` to public API to assist with date parsing.
- Corrects capitalisation of "DESFire".

# Verification 

- Scan Opal cards, should show correct balance and last activity
- Open a previous scan of an Opal card, should show up in the `Info` screen
- Scan a non-Opal DESFire card, that should not show anything about Opal
- Open a previous scan of a non-Opal DESFire card, that should also not show anything about Opal
- Repeat tests with different date / time formatting settings, that should follow system settings
- Check through different values that exercise formatting edge cases

Example dump of some slightly modified transit litter: [OpalLitter.nfc.txt](https://github.com/flipperdevices/flipperzero-firmware/files/11523840/OpalLitter.nfc.txt)

* Card number: `3085 2200 7856 2242`
* Balance: $-1.82
* Has automatic top-up enabled (modified)
* Weekly trips: 2
* Transaction number: 21
* Last transaction timestamp: 2015-07-27 at 18:24
* Last transaction mode: Rail / Metro
* Last transaction type: Tap off (distance fare)
* Checksum probably doesn't match due to modifications, but this doesn't check it.



# Checklist (For Reviewer)
